processor/metric/payload.go

```diff
@@ -0,0 +1,300 @@
+package metric
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/config"
+	m "github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/utility"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/monitoring"
+)
+
+var (
+	transformations = monitoring.NewInt(metricMetrics, "transformations")
+	processorEntry  = common.MapStr{"name": processorName, "event": docType}
+
+	errMissingKey = errors.New("key not found")
+	errNan        = errors.New("not a json number")
+)
+
+type Sample interface {
+	Name() string
+	Value() common.MapStr
+}
+
+func floatFrom(key string, m map[string]interface{}) (float64, error) {
+	v, ok := m[key]
+	if !ok {
+		return 0, errMissingKey
+	}
+	n, ok := v.(json.Number)
+	if !ok {
+		return 0, errNan
+	}
+	return n.Float64()
+}
+
+func stringPtrFrom(key string, m map[string]interface{}) *string {
+	if v, ok := m[key]; ok {
+		if s, ok := v.(string); ok {
+			return &s
+		}
+	}
+	return nil
+}
+
+type Counter struct {
+	name  string
+	count float64
+	unit  *string
+}
+
+func (c *Counter) Name() string {
+	return c.name
+}
+
+func (c *Counter) Value() common.MapStr {
+	v := common.MapStr{
+		"type":  "counter",
+		"value": c.count,
+	}
+	if c.unit != nil {
+		v["unit"] = c.unit
+	}
+	return v
+}
+
+type Gauge struct {
+	name  string
+	value float64
+	unit  *string
+}
+
+func (g *Gauge) Name() string {
+	return g.name
+}
+
+func (g *Gauge) Value() common.MapStr {
+	v := common.MapStr{
+		"type":  "gauge",
+		"value": g.value,
+	}
+	if g.unit != nil {
+		v["unit"] = g.unit
+	}
+	return v
+}
+
+type Summary struct {
+	name             string
+	unit             *string
+	count, sum       float64  // required
+	max, min, stddev *float64 // optional
+}
+
+func (s *Summary) Name() string {
+	return s.name
+}
+
+func (s *Summary) Value() common.MapStr {
+	v := common.MapStr{
+		"type":  "summary",
+		"count": s.count,
+		"sum":   s.sum,
+	}
+	if s.unit != nil {
```

Comment:
Why not using the `utilty.MapStrEnhancer` https://github.com/elastic/apm-server/blob/master/utility/map_str_enhancer.go#L10?